### ISSUE TITLE
Slow load time due to number of queries

### ISSUE THREAD

**Issue by kacperp94** [dpgaspar/Flask-AppBuilder#875]
### Environment

Flask-Appbuilder version:
Flask-AppBuilder==1.12.2
pip freeze output:

- alabaster==0.7.12
- alembic==1.0.5
- Babel==2.6.0
- blinker==1.4
- blockdiag==1.5.4
- certifi==2018.11.29
- chardet==3.0.4
- Click==7.0
- colorama==0.4.1
- decorator==4.3.0
- defusedxml==0.5.0
- docutils==0.14
- Flask==1.0.2
- Flask-AppBuilder==1.12.2
- Flask-Babel==0.12.2
- Flask-DebugToolbar==0.10.1
- Flask-Login==0.4.1
- Flask-Migrate==2.3.1
- Flask-MySQLdb==0.2.0
- Flask-OpenID==1.2.5
- Flask-SQLAlchemy==2.3.2
- Flask-WTF==0.14.2
- funcparserlib==0.3.6
- idna==2.8
- imagesize==1.1.0
- infinity==1.4
- intervals==0.8.1
- itsdangerous==1.1.0
- JayDeBeApi==1.1.1
- Jinja2==2.10
- JPype1==0.6.3
- Mako==1.0.7
- MarkupSafe==1.1.0
- mysqlclient==1.3.14
- packaging==18.0
- Pillow==5.3.0
- Pygments==2.3.0
- pyparsing==2.3.0
- python-dateutil==2.7.5
- python-editor==1.0.3
- python3-openid==3.1.0
- pytz==2018.7
- requests==2.21.0
- six==1.12.0
- snowballstemmer==1.2.1
- Sphinx==1.8.2
- sphinx-rtd-theme==0.4.2
- sphinxcontrib-blockdiag==1.5.5
- sphinxcontrib-websupport==1.1.0
- SQLAlchemy==1.2.14
- SQLAlchemy-Utils==0.33.9
- urllib3==1.24.1
- validators==0.12.3
- webcolors==1.8.1
- Werkzeug==0.14.1
- WTForms==2.2.1
- WTForms-Alchemy==0.16.8
- WTForms-Components==0.10.3

### Describe the expected results

Tell us what should happen.

I would expect pages to load in less than 1.5-2s. It seems like FAB is querying db ~60 times with roughly 90% of these being the same query. Since we are using external database (MariaDB) each lookup takes ~45-60ms and isn't done in parallel.

### Describe the actual results

Pages are taking ~4s+ to load and there are what seems to be multiple queries that are the same. 
<img width="832" alt="screen shot 2018-12-18 at 8 52 31 am 1" src="https://user-images.githubusercontent.com/10733732/50162720-9b9a8d00-02a4-11e9-9543-a72a200d570d.png">
I just want to know if there is a way to reduce the number of queries and consequently speed up the website.

### Steps to reproduce
The same issue happens with the new version of apache-airflow we use (old version using flask-admin is fast), so I would assume to reproduce the issue it should be sufficient to create a new flask-appbuilder project and set it up with external database.

**Comment by dpgaspar** [comment]
Hi @kacperp94,

These queries are not repeated, but this could be optimised. That being said.
40-60ms seems a lot to me, for an index query. Assuming connection pooling on flask-sqlalchemy you could discard network latency (but this should be checked), but there is an enormous array of possible performance problems, including network, disk io, shared db and whatnot.

Have you figure this one out already?

**Comment by kacperp94** [comment]
Hi @dpgaspar, I managed to resolve the issue.
First I tried using 2 databases, an instance of local SQLite db for all the `ab` models and external cloud db for all the actual data. That solved the load time issue, but since we wanted to have the user data available in the external database we had to find another solution.
What ultimately solved the issue was keeping our cloud db instance and server in the same location. 
So the issue was caused by network latency between server and the cloud db.
